bobgit/add.go

```diff
@@ -0,0 +1,116 @@
+package bobgit
+
+import (
+	"bufio"
+	"bytes"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/Benchkram/bob/bobgit/add"
+	"github.com/Benchkram/bob/pkg/bobutil"
+	"github.com/Benchkram/bob/pkg/cmdutil"
+	"github.com/Benchkram/bob/pkg/usererror"
+	"github.com/Benchkram/errz"
+)
+
+// Status executes `git add` in all repositories
+// first level repositories found inside a .bob filtree.
+// run git add commands by travsersing all the repositories
+// inside the bob workspace. if target is provided "." or  ""
+// it runs `git add .` in all repos, else run `git add ${relativeTargetPath}`
+// only on the selected repos depending on the target path
+func Add(target string) (err error) {
+	defer errz.Recover(&err)
+
+	if strings.HasPrefix(target, "/") {
+		return usererror.Wrap(ErrOutsideBobWorkspace)
+	}
+
+	bobRoot, err := bobutil.FindBobRoot()
+	errz.Fatal(err)
+
+	target, err = convertTargetPathRelativeToRoot(bobRoot, target)
+	if err != nil {
+		return usererror.Wrap(err)
+	}
+
+	err = os.Chdir(bobRoot)
+	errz.Fatal(err)
+
+	// Assure toplevel is a git repo
+	isGit, err := isGitRepo(bobRoot)
+	errz.Fatal(err)
+	if !isGit {
+		return usererror.Wrap(ErrCouldNotFindGitDir)
+	}
+
+	at := add.NewTarget(target)
+
+	// search for git repos inside bobRoot/.
+	allRepos, err := getAllRepos(bobRoot)
+	errz.Fatal(err)
+
+	filteredRepos := at.SelectReposByTarget(allRepos)
+
+	for _, name := range filteredRepos {
+		thistarget, err := at.GetRelativeTarget(name)
+		errz.Fatal(err)
+
+		if name == "." {
+			name = strings.TrimSuffix(name, ".")
+		}
+
+		output, err := cmdutil.GitAddDry(name, thistarget)
+		if err != nil {
+			return usererror.Wrapm(err, "Failed to Add files to git.")
+		}
+
+		filenames := parseAddDryOutput(output)
+
+		if len(filenames) > 0 {
+			err = cmdutil.GitAdd(name, thistarget)
+			if err != nil {
+				return usererror.Wrapm(err, "Failed to Add files to git.")
+			}
+		}
+	}
+
+	return nil
+}
+
+func parseAddDryOutput(buf []byte) (_ []string) {
```

Comment:
comment added, but not sure about the examples. This function just parses the output from add dry run and returns the filenames. If it runs fine for all pathspecs, then it will run the final `git add` command or else rejects the whole command for every file.